rfcs/0004-transaction.md

```diff
@@ -0,0 +1,258 @@
+2. Transaction 2 first checks the total amount of balance in A and B, and then transfers $450 from account B to account D.
+
+In this case, if the two transactions are executed concurrently, the total amount of balance in A and B might be less than $200, which violates the business rule. There's no Write-Write conflict in the two transactions, but the Read-Write conflict is not detected by Snapshot Isolation.
+
+To mitigate this anomaly, we could choose:
+
+1. use Serializable Snapshot Isolation (SSI) level: increase the isolation level, and no longer allow the write skew anomaly to occur.
+2. use `get_for_update()` to synchronize the critical read operations and write operations in the SI transaction.
+
+(1) is the approach that Badger takes, it supported a lightweight SSI transaction implementation.
+
+(2) is the approach that RocksDB takes, it does not built in SSI support, but it provides a `GetForUpdate()` method to lock the key before the write operation.
+
+In my understanding, the pros of the RocksDB approach is mostly easy to implement and efficient, and free of floating garbage (when compares with innodb's UNDO logs approach and the Postgres's multi version rows approach). Also, in most of the time, users can make the decision by themselves about when to use `GetForUpdate()` to make the critical concurrency control correct.
+
+However, it may produce massive writes on read intensive workloads when you hope to keep the read & write operations consistent, because `GetForUpdate()` simply transforms an read operation into a write operation, rewriting the key/value pair into WAL again. Also, some researches found that manual `GetForUpdate()` often introduces unexpected hard to diagnosis bugs, as described in the "2.2 Why Serializability" section of the [Serializable Snapshot Isolation in PostgreSQL](https://arxiv.org/pdf/1208.4179) paper:
+
+> Given the existence of these techniques, one might question the
+> need to provide serializable isolation in the database: shouldn’t users
+> just program their applications to handle the lower isolation level?
+> (We have often been asked this question.) Our view is that providing
+> serializability in the database is an important simplification for
+> application developers, because concurrency issues are notoriously
+> difficult to deal with. Indeed, SI anomalies have been discovered
+> in real-world applications [14]. The analysis required to identify
+> potential anomalies (or prove that none exist) is complex and is
+> likely beyond the reach of many users. In contrast, serializable
+> transactions offer simple semantics: users can treat their transactions
+> as though they were running in isolation.
+
+As above, I think it's better to finally have SSI support in the transaction feature, but thanks to the simplicity of the RocksDB approach, we could consider to implement the RocksDB approach (Snapshot Isolation + `GetForUpdate()`) first, and then implement the SSI support after it.
```

Comment:
> I think it's better to finally have SSI support in the transaction feature

👍 I agree. I am partial to skipping `GetForUpdate` and doing first SI then SSI afterwards, as you suggest below.